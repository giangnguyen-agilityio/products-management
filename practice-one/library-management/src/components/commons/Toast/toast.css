.toast {
  border-radius: 5px;
  position: fixed;
  display: flex;
  flex-direction: column;
  justify-content: center;
  top: 20px;
  z-index: 2;
  padding: 15px 15px 0px;
  right: 0;
}

.toast-content {
  display: flex;
  align-items: center;
  justify-content: center;
  font-size: 20px;
  font-family: var(--primary-font);
  color: var(--secondary-text-color);
  gap: 10px;
}

.toast.success {
  background-color: var(--primary-color);
}

.toast.failure {
  background-color: var(--danger-color);
}

.slide-in {
  opacity: 1;
  animation: slideIn 0.5s ease-in-out;
}

@keyframes slideIn {
  from {
    transform: translateX(100%);
  }
  to {
    transform: translateX(0);
  }
}

@keyframes slideOut {
  from {
    transform: translateX(0);
  }
  to {
    transform: translateX(100%);
  }
}

.border-bottom {
  height: 4px;
  background-color: var(--secondary-text-color);
  animation: borderEffect 5s ease-in-out;
  margin: 10px 0;
}

@keyframes borderEffect {
  from {
    width: 0%;
  }
  to {
    width: 100%;
  }
}
