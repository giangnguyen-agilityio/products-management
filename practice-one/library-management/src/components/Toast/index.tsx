import React, {useState, useEffect} from 'react'

// Importing the Icons form the React-icons library
import {AiOutlineCheckCircle, AiOutlineExclamationCircle} from 'react-icons/ai'

// Importing the CSS file for styling
import './toast.css'

// Define the props for the Toast component
interface ToastProps {
  message: string
  duration: number
  status: boolean
}

const Toast: React.FC<ToastProps> = props => {
  // Destructure the props
  const {message, duration, status} = props

  // State to manage the visibility of the toast
  const [isVisible, setIsVisible] = useState(false)

  useEffect(() => {
    // Set the visibility to true when the component is mounted
    setIsVisible(true)

    // Set a timer to hide the toast after the specified duration
    const timer = setTimeout(() => {
      setIsVisible(false)
    }, duration)

    // Clean up the timer when the component is unmounted or duration changes
    return () => {
      clearTimeout(timer)
    }
  }, [duration])

  // If the toast is not visible, don't render anything
  if (!isVisible) {
    return null
  }

  // Render the toast component with appropriate styling and content
  return (
    <div
      className={`toast ${status ? 'success' : 'failure'} ${
        isVisible ? 'slide-in' : 'slide-out'
      }`}
    >
      <div className="toast-content">
        {/* Render success or failure icon based on the status prop */}
        {status ? <AiOutlineCheckCircle /> : <AiOutlineExclamationCircle />}
        {message}
      </div>
      <div className="border-bottom"></div>
    </div>
  )
}

export default Toast
