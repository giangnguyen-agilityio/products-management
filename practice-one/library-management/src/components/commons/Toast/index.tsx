import React, {useState, useEffect} from 'react'
import {AiOutlineCheckCircle, AiOutlineExclamationCircle} from 'react-icons/ai'
import './toast.css'

interface ToastProps {
  message: string
  duration: number
  status: boolean
}

const Toast: React.FC<ToastProps> = props => {
  const {message, duration, status} = props

  // State to manage the visibility of the toast
  const [isVisible, setIsVisible] = useState(false)

  useEffect(() => {
    setIsVisible(true)
    const timer = setTimeout(() => {
      setIsVisible(false)
    }, duration)

    // Clean up the timer when the component is unmounted or duration changes
    return () => {
      clearTimeout(timer)
    }
  }, [duration])

  if (!isVisible) {
    return null
  }

  return (
    <div
      className={`toast ${status ? 'success' : 'failure'} ${
        isVisible && 'slide-in'
      }`}
    >
      <div className="toast-content">
        {/* Render success or failure icon based on the status prop */}
        {status ? <AiOutlineCheckCircle /> : <AiOutlineExclamationCircle />}
        {message}
      </div>
      <div className="border-bottom"></div>
    </div>
  )
}

export default Toast
