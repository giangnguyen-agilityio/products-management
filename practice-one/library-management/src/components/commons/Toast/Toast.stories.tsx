import {type StoryFn, type Meta} from '@storybook/react'

import Toast from './index'

export default {
  title: 'Example/Toast',
  component: Toast,
} as Meta<typeof Toast>

const Template: StoryFn<typeof Toast> = args => <Toast {...args} />

export const Default = Template.bind({})
Default.args = {
  duration: 5000,
  message: 'Book added successfully',
  status: true,
}

export const Secondary = Template.bind({})
Secondary.args = {
  duration: 5000,
  message: 'Failed to add a new book',
  status: false,
}
