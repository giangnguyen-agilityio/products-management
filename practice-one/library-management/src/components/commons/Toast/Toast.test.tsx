import React from 'react'
import {act, render, waitFor} from '@testing-library/react'

import Toast from '.'

// Mock setTimeout to control timer behavior
jest.useFakeTimers()

describe('Toast component', () => {
  it('should renders with success status', () => {
    const props = {
      message: 'Success Message',
      duration: 3000,
      status: true,
    }

    const {container, getByText} = render(<Toast {...props} />)
    const svgElement = container.querySelector('svg')

    expect(getByText('Success Message')).toBeInTheDocument()
    expect(svgElement).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })

  it('should renders with failure status', () => {
    const props = {
      message: 'Failure Message',
      duration: 3000,
      status: false,
    }

    const {container, getByText} = render(<Toast {...props} />)
    const svgElement = container.querySelector('svg')

    expect(getByText('Failure Message')).toBeInTheDocument()
    expect(svgElement).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })

  it('should not apply the "slide-in" class when isVisible is false', async () => {
    const props = {
      message: 'Message',
      duration: 3000,
      status: true,
    }

    const {container} = render(<Toast {...props} />)

    // Fast-forward time to simulate the duration
    // Advance timers by the duration to trigger the useEffect cleanup function
    act(() => {
      jest.advanceTimersByTime(props.duration)
    })

    // Wait for the component to re-render after the state change
    await waitFor(() => {
      // After the duration: slide-in class should not be applied
      const toastElement = container.querySelector('.toast')
      expect(toastElement).toBeNull()
      expect(toastElement).toMatchSnapshot()
    })
  })
})
