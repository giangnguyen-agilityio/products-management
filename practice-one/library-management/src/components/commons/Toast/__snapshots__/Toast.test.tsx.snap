// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toast component should not apply the "slide-in" class when isVisible is false 1`] = `null`;

exports[`Toast component should renders with failure status 1`] = `
<div>
  <div
    class="toast failure slide-in"
  >
    <div
      class="toast-content"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 1024 1024"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
        />
        <path
          d="M464 688a48 48 0 1 0 96 0 48 48 0 1 0-96 0zm24-112h48c4.4 0 8-3.6 8-8V296c0-4.4-3.6-8-8-8h-48c-4.4 0-8 3.6-8 8v272c0 4.4 3.6 8 8 8z"
        />
      </svg>
      Failure Message
    </div>
    <div
      class="border-bottom"
    />
  </div>
</div>
`;

exports[`Toast component should renders with success status 1`] = `
<div>
  <div
    class="toast success slide-in"
  >
    <div
      class="toast-content"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 1024 1024"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M699 353h-46.9c-10.2 0-19.9 4.9-25.9 13.3L469 584.3l-71.2-98.8c-6-8.3-15.6-13.3-25.9-13.3H325c-6.5 0-10.3 7.4-6.5 12.7l124.6 172.8a31.8 31.8 0 0 0 51.7 0l210.6-292c3.9-5.3.1-12.7-6.4-12.7z"
        />
        <path
          d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
        />
      </svg>
      Success Message
    </div>
    <div
      class="border-bottom"
    />
  </div>
</div>
`;
